transport_nantes/photo/models.py

```diff
@@ -63,6 +69,81 @@ class CategoryType(models.TextChoices):
         verbose_name="Date de soumission",
         auto_now_add=True,
     )
+    # Photo entries are pending (null) until they are accepted (True)
+    # or rejected (False)
+    accepted = models.BooleanField(null=True, blank=True, default=None)
+    # markdown text field where we'll add some interesting information
+    # about pedestrian issues.
+    pedestrian_issues_md = models.TextField(
+        blank=True, null=True,
+        verbose_name="Problèmes rencontrés par les piétons")
+    # markdown text field where we'll add information about the submitter
+    # and how the photo was taken.
+    submitter_info_md = models.TextField(
+        blank=True, null=True,
+        verbose_name="Informations sur le photographe")
+
+    # sha1 is used to generate a unique URL for each photo entry while keeping
+    # the ID private.
+    sha1_name = models.CharField(max_length=200, unique=True, editable=False)
 
     def __str__(self):
         return f"{self.user.username} - {self.category}"
+
+    def save(self, *args, **kwargs):
+        if not self.sha1_name:
+            self.sha1_name = self.generate_entry_sha1()
+
+        try:
+            super().save(*args, **kwargs)
+        except IntegrityError:
+            logger.warning(f"PhotoEntry {self.sha1_name} already exists")
+            self.sha1_name = self.generate_entry_sha1()
+            logger.warning(f"New session SHA1: {self.sha1_name}")
+            self.save(*args, **kwargs)
+
+    def generate_entry_sha1(self):
+        """Compute a sha1 from now() and the users's ID"""
+        import hashlib
+        import time
+        return hashlib.sha1(
+            (
+                str(time.monotonic_ns()) + "|" + str(self.user.id)
+            ).encode('utf-8')
+        ).hexdigest()
+
+
+class Vote(models.Model):
+    """
+    Represents a vote for a photo entry.
+
+    The same user can vote several times on the same entry, but only the last
+    vote is taken into account.
+
+    Field descriptions:
+    - user: the user who voted, can be none if user is anonymous
+    - tn_session_id : the tn session we set with SessionMiddleware. It's a
+    fallback mechanism in case the user is anonymous.
```

Comment:
I'm not convinced that tn_session_id is a fallback mechanism.  It's really a primary way of tracking visitor data without needing authenticated users and so respecting privacy.  We never allow ourselves to make persistent conclusions about users based on their non-logged-in behaviour (even if we do sometimes need to map the two together, ephemerally, as here).

Reply:
How would you describe its use then ? I used it as a way to identify a user that isn't logged in, it's kind of a default in absence of user account.

Reply:
Yes, but it's not a fall-back is all.  It's a first class identifier of site visitors.